internals/overlord/pairingstate/manager.go

```diff
@@ -0,0 +1,277 @@
+	return c.Mode == ModeUnset
+}
+
+func (c *PairingConfig) Combine(other *PairingConfig) {
+	if other.Mode != ModeUnset {
+		c.Mode = other.Mode
+	}
+}
+
+type PairingManager struct {
+	state  *state.State
+	mu     sync.Mutex
+	config *PairingConfig
+	open   bool
+	// timer controls the duration of the pairing window.
+	timer Timer
+	// cancelTimerFunc cancels a pairing window if enabled.
+	cancelTimerFunc context.CancelFunc
+}
+
+func NewManager(state *state.State) *PairingManager {
+	m := &PairingManager{
+		state: state,
+		config: &PairingConfig{
+			Mode: ModeDisabled,
+		},
+	}
+	return m
+}
+
+// PlanChanged informs the pairing manager that the plan has been updated.
+func (m *PairingManager) PlanChanged(update *plan.Plan) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	newConfig := update.Sections[PairingField].(*PairingConfig)
+
+	// If the mode changed, force the pairing window to be reopened
+	// taking the new config into account.
+	if m.config.Mode != newConfig.Mode {
+		m.open = false
+		m.stopTimer()
+	}
+	m.config = newConfig
+}
+
+// Ensure implements StateManager.Ensure.
+func (m *PairingManager) Ensure() error {
+	return nil
+}
+
+// PairMTLS adds a client identity with admin permissions to the identity
+// subsystem. A pairing request always closes the pairing window.
+func (m *PairingManager) PairMTLS(clientCert *x509.Certificate) error {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if !m.open {
+		return errors.New("cannot pair client: pairing is not open")
+	}
+
+	// Any success or failure should always close the pairing window.
+	defer func() {
+		m.open = false
+		m.stopTimer()
+	}()
+
+	m.state.Lock()
+	defer m.state.Unlock()
+
+	existingIdentities := m.state.Identities()
+
+	for _, identity := range existingIdentities {
+		if identity.Cert != nil && identity.Cert.X509 != nil {
+			if identity.Cert.X509.Equal(clientCert) {
+				return errors.New("cannot pair client: identity already paired")
+			}
+		}
+	}
+
+	username := m.generateUniqueUsername(existingIdentities)
+
+	newIdentity := &state.Identity{
+		Access: state.AdminAccess,
+		Cert:   &state.CertIdentity{X509: clientCert},
+	}
+
+	err := m.state.AddIdentities(map[string]*state.Identity{
+		username: newIdentity,
+	})
+	if err != nil {
+		return fmt.Errorf("cannot add identity: %w", err)
+	}
+
+	m.state.SetIsPaired()
+
+	return nil
+}
+
+// generateUniqueUsername generates a unique username following the pattern "user-x"
+// where x starts at 1 and monotonically increments. Users names not following this
+// pattern will simply be ignored.
+func (m *PairingManager) generateUniqueUsername(existingIdentities map[string]*state.Identity) string {
+	maxUserNumber := 0
+
+	for name := range existingIdentities {
+		if strings.HasPrefix(name, "user-") {
+			numberStr := strings.TrimPrefix(name, "user-")
+			if number, err := strconv.Atoi(numberStr); err == nil && number > maxUserNumber {
+				maxUserNumber = number
+			}
+		}
+	}
+
+	return fmt.Sprintf("user-%d", maxUserNumber+1)
```

Comment:
I know this is super painful, but how would this behave if an identity is added through `add-identities` locally with an user like `user-2147483647`?
Could we forbid the `user-` prefix entirely?